simpleviewinc/keg-herkin
# Display definition meta-data

## Context
* With Step Definitions, it's not always clear what the **value** should be for dynamic variables 
* For example in cases whe the dynamic variable is of type `{string}`
  * This could literally be anything, but the Step Definition expects it to be something very specific
  * Yet, it's up to the user to try and guess what it should be
* [Parkin](https://github.com/lancetipton/parkin) allows passing metadata with a Definition when it's registered
  * If we add a way to display the metadata, we can
    * Remove ambiguity
    * Make dynamic variable requirements more clear
    * Help the user enter the correct value for a definition

## Goal
* Allow Step Definition meta-data to be displayed
* Fix lower-case issue for Step Definition match-text

## Updates
* `repos/tap/assets/icons/*`
  * Added some icons for the UI
* `repos/tap/components/definition/definitionList.js`
  * Updated list to auto-(open/close) each group relative to the others
    * When one group opens, all others should be closed
  * Moved header-toggle logic to the `definitionList`
* `repos/tap/components/definition/definitionListItem.js`
  * Rewrote the definition list items
  * Added ability to toggle open meta-data when the item is clicked on
  * Updated the definition list items title formatting
    * Fixed capital letters being force to lowercase
* `repos/tap/components/definition/definitionMeta.js`
  * Added component to render meta-data of a definition when it's defined
* `repos/tap/components/list/listItem.js`
  * Updated to use `React.memo`
  * Moved render prop up to the parent component
  * Now listItem never gets called if `renderItem` method is passed
  * Added exports of child components on the ListItem component
* `repos/tap/components/list/listItemAction.js`
  * Fixed some issues with feedback not being shown properly
* `repos/tap/components/list/simpleList.js`
  * Updated ListItem render to call the `renderItem` method here
  * Added ability to manage header toggle outside the SimpleList component
* `repos/tap/components/sidebar/sidebar.js`
  * Cleaned up some spelling
  * Added a stupid-fun animation to the sidebar toggle
* `repos/tap/containers/rootContainer.js`
  * Added sidebar animation props
* `repos/tap/theme/components/definitions/definitionList.js`
  * Styles updates for the definition list item, and meta-data
* `repos/testUtils/steps/*`
  * Updated a couple of the core step-definitions to include meta-data

## Testing

**Setup**
* Pull this PR => `keg herkin && keg pr 34`
* Start keg-herkin => `keg herkin start --from keg-herkin:display-meta-data --config /path/to/config`
* Once started navigate to http://herkin-display-meta-data.local.kegdev.xyz/
* Once loaded, select any test feature file, does not matter which one

**Test 1**
* On the Editors tab, on the right is the definitions list
* By default the `All` section is toggled open
* Scroll, down and toggle one of the other sections ( Given | When | Then ) by clicking on it
* Ensure the `All` section closes automatically, as the section you clicked on opens up


**Test 2**
* Update a core definition's match text to include capital letters
  * Save the changes and reload the definitions
    * Find the updated definition in the definitions list
    * Ensure it allows capital letters
    * [Looks something list this](https://prnt.sc/1176km7)

**Test 3**
* A few of the definitions have meta data added to them 
* [Looks list this](http://snpy.in/pVG7uM)
* In the definitions list, click on any of the Definitions that have a point next to them
  * This should slide-toggle the metadata
  * Ensure the metadata is displayed properly
* Click on the Definitions text again to hide the metadata
  * Ensure the metadata slide-toggles to hidden
* Hover over a definition and ensure the `Copy to Clipboard` action displays
  * Click on the `Copy to Clipboard` action
  * Ensure the match text is still copied to the clipboard
  * Ensure you can paste the match text in the feature editor
